// zotero-plugin-scaffold/dist/shared/zotero-plugin-scaffold.v8dJZZ-v.mjs
import { readFile, writeFile } from 'node:fs/promises';
import { glob } from 'tinyglobby';

function dateFormat(fmt, date) {
  let ret;
  const opt = {
    "Y+": date.getFullYear().toString(),
    "m+": (date.getMonth() + 1).toString(),
    "d+": date.getDate().toString(),
    "H+": date.getHours().toString(),
    "M+": date.getMinutes().toString(),
    "S+": date.getSeconds().toString()
  };
  for (const k in opt) {
    ret = new RegExp(`(${k})`).exec(fmt);
    if (ret) {
      fmt = fmt.replace(
        ret[1],
        ret[1].length === 1 ? opt[k] : opt[k].padStart(ret[1].length, "0")
      );
    }
  }
  return fmt;
}
function toArray(value) {
  return Array.isArray(value) ? value : [value];
}
function template(str, data, regex = /\{\{(.+?)\}\}/g) {
  return Array.from(str.matchAll(regex)).reduce((acc, match) => {
    return acc.replace(match[0], data[match[1]]);
  }, str);
}
function parseRepoUrl(url) {
  if (!url)
    throw new Error("Parse repository URL failed.");
  const match = url.match(/:\/\/.+com\/([^/]+)\/([^.]+)\.git$/);
  if (!match) {
    throw new Error("Parse repository URL failed.");
  }
  const [, owner, repo] = match;
  return { owner, repo };
}
function replace(contents, from, to) {
  const froms = Array.isArray(from) ? from : [from];
  const tos = Array.isArray(to) ? to : Array.from({ length: froms.length }, () => to);
  if (froms.length !== tos.length) {
    throw new Error("The lengths of 'from' and 'to' must be equal");
  }
  return froms.reduce((result, pattern, index) => result.replace(pattern, tos[index]), contents);
}
async function replaceInFile({ files, from, to, isGlob = true }) {
  const paths = isGlob ? await glob(files) : toArray(files);
  await Promise.all(paths.map(async (path) => {
    const contents = await readFile(path, "utf-8");
    const newContents = replace(contents, from, to);
    if (contents !== newContents)
      await writeFile(path, newContents);
  }));
}

export { replaceInFile as a, toArray as b, dateFormat as d, parseRepoUrl as p, replace as r, template as t };
