// zotero-plugin-scaffold/dist/shared/zotero-plugin-scaffold.C-SXQX4J.mjs
import { loadConfig as loadConfig$1 } from 'c12';
import { isPlainObject, kebabCase, mapValues, toMerged, escapeRegExp, debounce, delay } from 'es-toolkit';
import { readJsonSync, move, outputFile, emptyDir, copy, readJSON, outputJSON, writeJson, pathExists, ensureDir, remove } from 'fs-extra/esm';
import { createHooks } from 'hookable';
import process from 'node:process';
import readline from 'node:readline';
import util from 'node:util';
import { isDebug, isCI, isWindows, isMacOS, isLinux } from 'std-env';
import { p as parseRepoUrl, d as dateFormat, t as template, b as toArray, a as replaceInFile } from './zotero-plugin-scaffold.v8dJZZ-v.mjs';
import { dirname, basename, join, resolve, extname } from 'node:path';
import AdmZip from 'adm-zip';
import { build, context } from 'esbuild';
import { glob, globSync } from 'tinyglobby';
import { createHash } from 'node:crypto';
import { createReadStream, existsSync } from 'node:fs';
import { readFile, writeFile, stat } from 'node:fs/promises';
import { parse, serialize, Transformer } from '@fluent/syntax';
import { parseSync, printSync } from '@swc/core';
import { execSync, spawn } from 'node:child_process';
import { versionBump, ProgressEvent } from 'bumpp';
import { Octokit } from 'octokit';
import chokidar from 'chokidar';
import net from 'node:net';
import { Buffer } from 'node:buffer';
import domain from 'node:domain';
import EventEmitter from 'node:events';
import { Xvfb } from 'xvfb-ts';
import http from 'node:http';

const factory = (...formats) =>
  new Proxy(util.styleText.bind(util, formats), {
    get: (_, format) =>
      util.inspect.colors[format] && factory(...formats, format),
  });

const styleText = factory();

var LOG_LEVEL = /* @__PURE__ */ ((LOG_LEVEL2) => {
  LOG_LEVEL2[LOG_LEVEL2["TRACE"] = 0] = "TRACE";
  LOG_LEVEL2[LOG_LEVEL2["DEBUG"] = 1] = "DEBUG";
  LOG_LEVEL2[LOG_LEVEL2["INFO"] = 2] = "INFO";
  LOG_LEVEL2[LOG_LEVEL2["WARN"] = 3] = "WARN";
  LOG_LEVEL2[LOG_LEVEL2["ERROR"] = 4] = "ERROR";
  return LOG_LEVEL2;
})(LOG_LEVEL || {});
const SYMBOLS = {
  SUCCESS: styleText.green("\u2714"),
  INFO: styleText.blue("\u2139"),
  FAIL: styleText.red("\u2716"),
  TIP: styleText.blue("\u2192"),
  ERROR: styleText.bgRed(" ERROR "),
  WARN: styleText.bgYellow(" WARN "),
  DEBUG: styleText.grey("\u2699"),
  NONE: ""
};
const DEFAULT_OPTIONS = {
  SPACE: 0,
  NEW_LINE: false
};
const LOG_METHODS_CONFIG = {
  error: {
    level: 4 /* ERROR */,
    symbol: SYMBOLS.ERROR,
    wrapNewLine: true
  },
  warn: {
    level: 3 /* WARN */,
    symbol: SYMBOLS.WARN,
    wrapNewLine: true
  },
  tip: {
    level: 2 /* INFO */,
    symbol: SYMBOLS.TIP
  },
  info: {
    level: 2 /* INFO */,
    symbol: SYMBOLS.INFO
  },
  debug: {
    level: 1 /* DEBUG */,
    symbol: SYMBOLS.DEBUG
  },
  success: {
    level: 2 /* INFO */,
    symbol: SYMBOLS.SUCCESS
  },
  fail: {
    level: 4 /* ERROR */,
    symbol: SYMBOLS.FAIL
  },
  log: {
    level: 2 /* INFO */,
    symbol: SYMBOLS.NONE
  }
};
class Logger {
  static instance;
  currentLogLevel;
  constructor(level) {
    this.currentLogLevel = this.determineLogLevel(level);
  }
  /**
   * Determine the appropriate log level
   */
  determineLogLevel(level) {
    if (isDebug)
      return 0 /* TRACE */;
    const envLevel = process.env.ZOTERO_PLUGIN_LOG_LEVEL;
    return envLevel ? LOG_LEVEL[envLevel] : level ?? 2 /* INFO */;
  }
  static getInstance() {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }
  setLogLevel(level) {
    this.currentLogLevel = LOG_LEVEL[level];
  }
  get level() {
    return this.currentLogLevel;
  }
  /**
   * Generic log formatting logic
   */
  formatContent(content) {
    if (typeof content === "string")
      return content;
    if (content instanceof Error)
      return this.formatError(content);
    if (isPlainObject(content))
      return JSON.stringify(content, null, 2);
    return String(content);
  }
  formatError(error) {
    return `${styleText.red(error.name)}: ${styleText.red(error.message)}
${error.stack}`;
  }
  /**
   * Core logging method
   */
  logInternal(content, config, options = {}) {
    if (this.currentLogLevel > config.level)
      return;
    const { space = DEFAULT_OPTIONS.SPACE, newLine = DEFAULT_OPTIONS.NEW_LINE } = options;
    const formattedContent = this.formatContent(content);
    const output = [
      "  ".repeat(space),
      config.symbol,
      formattedContent
    ].join(" ");
    if (config.wrapNewLine)
      this.newLine();
    console.log(output);
    if (newLine || config.wrapNewLine)
      this.newLine();
  }
  // Public API methods
  error(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.error, options);
  }
  warn(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.warn, options);
  }
  tip(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.tip, options);
  }
  info(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.info, options);
  }
  debug(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.debug, options);
  }
  success(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.success, options);
  }
  fail(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.fail, options);
  }
  ready(content) {
    this.logInternal(styleText.green(content), LOG_METHODS_CONFIG.success);
  }
  log(content, options) {
    this.logInternal(content, LOG_METHODS_CONFIG.log, options);
  }
  clear() {
    const blank = process.stdout.rows > 2 ? "\n".repeat(process.stdout.rows - 2) : "";
    console.log(blank);
    readline.cursorTo(process.stdout, 0, 0);
    readline.clearScreenDown(process.stdout);
  }
  newLine() {
    console.log();
  }
}
const logger = Logger.getInstance();

function defineConfig(userConfig) {
  return userConfig;
}
async function loadConfig(overrides) {
  const result = await loadConfig$1({
    name: "zotero-plugin",
    dotenv: true,
    packageJson: true,
    // eslint-disable-next-line ts/no-use-before-define
    defaults: getDefaultConfig(),
    overrides
  });
  return resolveConfig(result.config);
}
function resolveConfig(config) {
  logger.setLogLevel(config.logLevel);
  const pkgUser = readJsonSync("package.json", {
    encoding: "utf-8"
  });
  const { name, version } = pkgUser;
  const { owner, repo } = parseRepoUrl(pkgUser.repository?.url);
  config.name ||= name;
  config.id ||= config.name;
  config.namespace ||= config.name;
  config.xpiName ||= kebabCase(config.name);
  config.build.prefs.prefix ||= pkgUser.config.prefsPrefix || `extensions.${config.namespace}`;
  const isPreRelease = version.includes("-");
  const templateData = {
    owner,
    repo,
    version,
    isPreRelease,
    updateJson: isPreRelease ? "update-beta.json" : "update.json",
    xpiName: config.xpiName,
    buildTime: dateFormat("YYYY-mm-dd HH:MM:SS", /* @__PURE__ */ new Date())
  };
  config.updateURL = template(config.updateURL, templateData);
  config.xpiDownloadLink = template(config.xpiDownloadLink, templateData);
  config.build.define = mapValues(config.build.define, (v) => template(v, templateData));
  config.release.github.repository = template(config.release.github.repository, templateData);
  const hooks = createHooks();
  hooks.addHooks(config.build.hooks);
  hooks.addHooks(config.server.hooks);
  hooks.addHooks(config.release.hooks);
  hooks.addHooks(config.test.hooks);
  const ctx = {
    ...config,
    pkgUser,
    version,
    hooks,
    logger,
    templateData
  };
  return ctx;
}
const defaultConfig = {
  source: "src",
  dist: ".scaffold/build",
  name: "",
  id: "",
  namespace: "",
  xpiName: "",
  xpiDownloadLink: "https://github.com/{{owner}}/{{repo}}/releases/download/v{{version}}/{{xpiName}}.xpi",
  updateURL: "https://github.com/{{owner}}/{{repo}}/releases/download/release/{{updateJson}}",
  build: {
    assets: "addon/**/*.*",
    define: {},
    fluent: {
      prefixFluentMessages: true,
      prefixLocaleFiles: true,
      ignore: [],
      dts: "typings/i10n.d.ts"
    },
    prefs: {
      prefix: "",
      prefixPrefKeys: true,
      dts: "typings/prefs.d.ts"
    },
    esbuildOptions: [],
    makeManifest: {
      enable: true,
      template: {
        manifest_version: 2,
        name: "",
        version: "",
        applications: {
          zotero: {
            id: "",
            update_url: ""
          }
        }
      }
    },
    makeUpdateJson: {
      updates: [],
      hash: true
    },
    hooks: {}
  },
  server: {
    devtools: true,
    startArgs: [],
    asProxy: false,
    prebuild: true,
    // keepProfileChanges: true,
    createProfileIfMissing: true,
    hooks: {}
  },
  addonLint: {},
  release: {
    bumpp: {
      release: "prompt",
      preid: "beta",
      confirm: true,
      execute: "",
      all: false,
      commit: "chore(publish): release v%s",
      noVerify: false,
      tag: "v%s"
    },
    changelog: "",
    github: {
      enable: "ci",
      repository: "{{owner}}/{{repo}}",
      updater: "release",
      comment: false,
      releaseNote: (ctx) => {
        return ctx.release.changelog;
      }
    },
    hooks: {}
  },
  test: {
    entries: "test",
    prefs: {},
    mocha: {
      timeout: 1e4
    },
    abortOnFail: false,
    headless: false,
    startupDelay: 1e3,
    waitForPlugin: "() => true",
    watch: true,
    hooks: {}
  },
  logLevel: "INFO"
};
const getDefaultConfig = () => defaultConfig;

function generateHash(filePath, algorithm) {
  return new Promise((resolve, reject) => {
    const hash = createHash(algorithm);
    const stream = createReadStream(filePath);
    stream.on("data", (data) => {
      hash.update(data);
    });
    stream.on("end", () => {
      const fileHash = hash.digest("hex");
      resolve(`${algorithm}:${fileHash}`);
    });
    stream.on("error", (error) => {
      reject(error);
    });
  });
}

class Base {
  ctx;
  constructor(ctx) {
    this.ctx = ctx;
  }
  get logger() {
    return this.ctx.logger;
  }
}

async function buildLocale(dist, namespace, options) {
  const ignores = toArray(options.ignore);
  const localeNames = await getLocales(dist);
  const messageManager = new MessageManager(ignores);
  await Promise.all(localeNames.map(async (locale) => {
    const paths = await glob(`${dist}/addon/locale/${locale}/**/*.ftl`);
    await Promise.all(paths.map(async (path) => {
      const fm = new FluentManager();
      await fm.read(path);
      messageManager.addMessages(locale, fm.getMessages());
      if (options.prefixFluentMessages) {
        fm.prefixMessages(namespace);
        await fm.write(path);
      }
      if (options.prefixLocaleFiles) {
        const newPath = `${dirname(path)}/${namespace}-${basename(path)}`;
        await move(path, newPath);
        logger.debug(`Renamed FTL: ${path} \u2192 ${newPath}`);
      }
    }));
  }));
  const htmlPaths = await glob([`${dist}/addon/**/*.xhtml`, `${dist}/addon/**/*.html`]);
  await Promise.all(htmlPaths.map(async (htmlPath) => {
    const content = await readFile(htmlPath, "utf-8");
    const { processedContent, foundMessages } = processHTMLFile(
      content,
      namespace,
      messageManager.getFTLMessages(),
      ignores,
      htmlPath
    );
    messageManager.addMessages("html", foundMessages);
    if (options.prefixFluentMessages) {
      await writeFile(htmlPath, processedContent);
    }
  }));
  messageManager.validateMessages();
  if (options.dts) {
    const dtsContent = generateFluentDts([...messageManager.getFTLMessages()]);
    await outputFile(options.dts, dtsContent, "utf-8");
  }
}
async function getLocales(dist) {
  const localePaths = await glob(`${dist}/addon/locale/*`, { onlyDirectories: true });
  return localePaths.map((p) => basename(p));
}
class FluentManager {
  source;
  resource;
  messages = [];
  constructor() {
  }
  // Parse Fluent source into an AST and extract messages
  parse(source) {
    this.source = source;
    this.resource = parse(source, {});
  }
  // Read a file, parse its content, and extract messages
  async read(path) {
    const content = await readFile(path, "utf-8");
    this.parse(content);
  }
  // Extract message IDs from the parsed resource
  getMessages() {
    if (!this.resource) {
      throw new Error("Resource must be parsed first.");
    }
    this.messages.length = 0;
    this.messages.push(
      ...this.resource.body.filter((entry) => entry.type === "Message").map((message) => message.id.name)
    );
    return this.messages;
  }
  // Apply namespace prefix to message IDs in the resource
  prefixMessages(namespace) {
    if (!this.resource) {
      throw new Error("Resource must be parsed before applying prefix.");
    }
    new FluentTransformer(namespace).genericVisit(this.resource);
  }
  // Serialize the resource back into a string
  serialize() {
    if (!this.resource) {
      throw new Error("Resource not parsed. Cannot serialize.");
    }
    return serialize(this.resource, {});
  }
  // Write the serialized resource to a file
  async write(path) {
    const result = this.serialize();
    if (result !== this.source)
      await writeFile(path, this.serialize());
  }
}
class FluentTransformer extends Transformer {
  constructor(prefix) {
    super();
    this.prefix = prefix;
  }
  needsPrefix(name) {
    return !!this.prefix && !name.startsWith(this.prefix);
  }
  visitMessage(node) {
    if (this.needsPrefix(node.id.name)) {
      node.id.name = `${this.prefix}-${node.id.name}`;
    }
    return this.genericVisit(node);
  }
  visitMessageReference(node) {
    if (this.needsPrefix(node.id.name)) {
      node.id.name = `${this.prefix}-${node.id.name}`;
    }
    return this.genericVisit(node);
  }
}
class MessageManager {
  ftlMessages = /* @__PURE__ */ new Map();
  htmlMessages = /* @__PURE__ */ new Set();
  ignores;
  constructor(ignores) {
    this.ignores = ignores;
  }
  // Add a set of messages (FTL or HTML) for a specific locale or for HTML globally
  addMessages(target, messages) {
    if (target === "html") {
      messages.forEach((msg) => this.htmlMessages.add(msg));
    } else {
      let ftlLocaleMessages = this.ftlMessages.get(target);
      if (!ftlLocaleMessages) {
        ftlLocaleMessages = /* @__PURE__ */ new Set();
        this.ftlMessages.set(target, ftlLocaleMessages);
      }
      messages.forEach((msg) => ftlLocaleMessages.add(msg));
    }
  }
  validateMessages() {
    this.htmlMessages.forEach((msg) => {
      if (this.ignores.includes(msg))
        return;
      const missingLocales = [...this.ftlMessages.entries()].filter(([_, messages]) => !messages.has(msg)).map(([locale]) => locale);
      if (missingLocales.length > 0)
        logger.warn(`I10N id ${styleText.blue(msg)} missing in locale: ${missingLocales.join(", ")}`);
    });
  }
  getFTLMessages() {
    const allMessages = /* @__PURE__ */ new Set();
    this.ftlMessages.forEach((messages) => messages.forEach((msg) => allMessages.add(msg)));
    return allMessages;
  }
  // Get all FTL messages for a specific locale
  getFTLMessagesByLocale(locale) {
    return this.ftlMessages.get(locale) || /* @__PURE__ */ new Set();
  }
  // Get all HTML messages
  getHTMLMessages() {
    return this.htmlMessages;
  }
}
function processHTMLFile(content, namespace, allMessages, ignores, filePath) {
  const foundMessages = /* @__PURE__ */ new Set();
  const L10N_PATTERN = new RegExp(`(data-l10n-id)="((?!${namespace})\\S*)"`, "g");
  const processed = content.replace(L10N_PATTERN, (match, attr, id) => {
    foundMessages.add(id);
    if (ignores.includes(id)) {
      logger.debug(`Skipped ignored ID: ${styleText.blue(id)} in ${styleText.gray(filePath)}`);
      return match;
    }
    if (!allMessages.has(id)) {
      logger.warn(`I10N id ${styleText.blue(id)} in path ${styleText.gray(filePath)} does not exist in any locale, skip renaming it.`);
      return match;
    }
    return `${attr}="${namespace}-${id}"`;
  });
  return { processedContent: processed, foundMessages: [...foundMessages] };
}
function generateFluentDts(messages) {
  return `// Generated by zotero-plugin-scaffold
/* prettier-ignore */
/* eslint-disable */
// @ts-nocheck
export type FluentMessageId =
${Array.from(messages).sort().map((id) => `  | '${id}'`).join("\n")};
`;
}

function is32BitNumber(n) {
  return Number.isInteger(n) && n >= -2147483648 && n <= 2147483647;
}

class PrefsManager {
  namespace;
  prefs = {};
  constructor(namespace) {
    this.namespace = namespace;
  }
  /**
   * Parse Method 3 - Using AST
   */
  parse(content) {
    const _map = {};
    const ast = parseSync(content, { syntax: "ecmascript" });
    for (const node of ast.body) {
      if (node.type !== "ExpressionStatement" || node.expression.type !== "CallExpression" || node.expression.callee.type !== "Identifier" || node.expression.callee.value !== this.namespace || node.expression.arguments.length !== 2) {
        throw new Error("Invalid prefs.js file.");
      }
      const [arg1, arg2] = node.expression.arguments;
      if (arg1.expression.type !== "StringLiteral") {
        throw new Error("Invalid prefs.js file - unsupported key type.");
      }
      const key = arg1.expression.value.trim();
      let value;
      switch (arg2.expression.type) {
        // https://babeljs.io/docs/babel-parser#output
        case "StringLiteral":
        case "NumericLiteral":
        case "BooleanLiteral":
          value = arg2.expression.value;
          break;
        // https://github.com/estree/estree/blob/master/es5.md#unaryexpression
        // https://github.com/northword/zotero-plugin-scaffold/issues/98
        case "UnaryExpression":
          if (arg2.expression.argument.type !== "NumericLiteral")
            throw new Error("Invalid prefs.js file - unsupported value type.");
          if (arg2.expression.operator === "-")
            value = -arg2.expression.argument.value;
          else if (arg2.expression.operator === "+")
            value = arg2.expression.argument.value;
          else
            throw new Error("Invalid prefs.js file - unsupported value type.");
          break;
        case "TemplateLiteral":
          value = arg2.expression.quasis[0]?.cooked ?? "";
          break;
        default:
          throw new Error("Invalid prefs.js file - unsupported value type.");
      }
      _map[key] = value;
    }
    return _map;
  }
  /**
   * Parse Method 1 - Using RegExp
   * @deprecated
   */
  parseByRegExp(content) {
    const _map = {};
    const prefPattern = /^(pref|user_pref)\s*\(\s*["']([^"']+)["']\s*,\s*(.+)\s*,?\s*\)\s*;?$/gm;
    const matches = content.matchAll(prefPattern);
    for (const match of matches) {
      const key = match[2].trim();
      const value = match[3].trim();
      _map[key] = this.cleanValue(value);
    }
    return _map;
  }
  /**
   * Parse Method 2 - Using eval
   * @deprecated
   */
  // private parseByEval(content: string) {
  //   const _map: Prefs = {};
  //   // eslint-disable-next-line unused-imports/no-unused-vars
  //   const pref = (key: any, value: any) => {
  //     _map[key.trim()] = this.cleanValue(value.trim());
  //   };
  //   // eslint-disable-next-line no-eval
  //   eval(content);
  //   return _map;
  // }
  cleanValue(value) {
    if (value === "true")
      return true;
    else if (value === "false")
      return false;
    else if (!Number.isNaN(Number(value)))
      return Number(value);
    else if (value.match(/^["'](.*)["']$/))
      return value.replace(/^["'](.*)["']$/, "$1");
    else
      return value;
  }
  /**
   * Render Method 2 - Using swc
   */
  render() {
    const span = { start: 0, end: 0, ctxt: 0 };
    function getExpression(value) {
      switch (typeof value) {
        case "string":
          return {
            type: "StringLiteral",
            span,
            value
          };
        case "boolean":
          return {
            type: "BooleanLiteral",
            span,
            value
          };
        case "number":
          if (value < 0) {
            return {
              type: "UnaryExpression",
              span,
              operator: "-",
              argument: {
                type: "NumericLiteral",
                span,
                value: Math.abs(value)
              }
            };
          }
          return {
            type: "NumericLiteral",
            span,
            value
          };
        default:
          throw new Error(`Unsupported value type: ${typeof value}`);
      }
    }
    const ast = {
      type: "Module",
      span,
      // @ts-expect-error no raw property
      body: Object.entries(this.prefs).map(([key, value]) => ({
        type: "ExpressionStatement",
        span,
        expression: {
          type: "CallExpression",
          span,
          ctxt: 0,
          callee: {
            type: "Identifier",
            span,
            ctxt: 0,
            value: this.namespace,
            optional: false
          },
          arguments: [
            { expression: getExpression(key) },
            { expression: getExpression(value) }
          ]
        }
      }))
    };
    const { code } = printSync(ast);
    return code;
  }
  /**
   * Render Method 1 - Using string
   * @deprecated
   */
  renderByString() {
    return Object.entries(this.prefs).map(([key, value]) => {
      const _v = typeof value === "string" ? `"${value.replaceAll("\\", "\\\\").replaceAll('"', '\\"')}"` : value;
      return `${this.namespace}("${key}", ${_v});`;
    }).join("\n");
  }
  async read(path) {
    const content = await readFile(path, "utf-8");
    const map = this.parse(content);
    this.setPrefs(map);
  }
  async write(path) {
    const content = this.render();
    await outputFile(path, content, "utf-8");
    logger.debug("The prefs.js has been modified.");
  }
  setPref(key, value) {
    if (value === null || value === void 0) {
      if (key in this.prefs)
        delete this.prefs[key];
      return;
    }
    this.prefs[key] = value;
  }
  setPrefs(prefs) {
    Object.entries(prefs).forEach(([key, value]) => {
      this.setPref(key, value);
    });
  }
  getPref(key) {
    return this.prefs[key] ?? void 0;
  }
  getPrefs() {
    return this.prefs;
  }
  clearPrefs() {
    this.prefs = {};
  }
  getPrefsWithPrefix(prefix) {
    const _prefs = {};
    for (const pref in this.prefs) {
      if (pref.startsWith(prefix))
        _prefs[pref] = this.prefs[pref];
      else
        _prefs[`${prefix}.${pref}`] = this.prefs[pref];
    }
    return _prefs;
  }
  getPrefsWithoutPrefix(prefix) {
    const _prefs = {};
    for (const pref in this.prefs) {
      _prefs[pref.replace(`${prefix}.`, "")] = this.prefs[pref];
    }
    return _prefs;
  }
}

async function buildPrefs(dist, options) {
  const { dts, prefixPrefKeys, prefix } = options;
  if (!prefixPrefKeys && !dts)
    return;
  const prefsFilePath = join(dist, "addon", "prefs.js");
  if (!existsSync(prefsFilePath))
    return;
  const prefsManager = new PrefsManager("pref");
  await prefsManager.read(prefsFilePath);
  const prefsWithPrefix = prefsManager.getPrefsWithPrefix(prefix);
  const prefsWithoutPrefix = prefsManager.getPrefsWithoutPrefix(prefix);
  const prefs = prefsManager.getPrefs();
  Object.entries(prefs).forEach(([key, value]) => {
    if (typeof value === "number") {
      if (!is32BitNumber(value)) {
        logger.warn(`Pref key '${styleText.blue(key)}' is a number, but is more than 4 bytes, which can be problematic on some OS.`);
      }
    }
  });
  if (dts) {
    const dtsContent = renderPluginPrefsDts(prefsWithoutPrefix);
    await outputFile(dts, dtsContent, "utf-8");
  }
  if (prefixPrefKeys) {
    prefsManager.clearPrefs();
    prefsManager.setPrefs(prefsWithPrefix);
    await prefsManager.write(prefsFilePath);
  }
  if (prefixPrefKeys) {
    const HTML_PREFERENCE_PATTERN = /preference="(\S*)"/g;
    const xhtmlPaths = await glob(`${dist}/addon/**/*.xhtml`);
    await Promise.all(xhtmlPaths.map(async (path) => {
      let content = await readFile(path, "utf-8");
      const matchs = [...content.matchAll(HTML_PREFERENCE_PATTERN)];
      for (const match of matchs) {
        const [matched, key] = match;
        if (key.startsWith(prefix)) {
          logger.debug(`Pref key '${styleText.blue(key)}' is already starts with '${prefix}', skip prefixing it.`);
          continue;
        } else if (key.startsWith("extensions.")) {
          logger.warn(`Pref key '${styleText.blue(key)}' in ${styleText.gray(path)} starts with 'extensions.' but not '${styleText.blue(prefix)}', skip prefixing it.`);
          continue;
        } else if (!(key in prefsWithPrefix) && !(key in prefsWithoutPrefix)) {
          logger.warn(`Pref key '${styleText.blue(key)}' in ${styleText.gray(path)} is not found in prefs.js, skip prefixing it.`);
          continue;
        } else {
          const prefixed = `${prefix}.${key}`;
          logger.debug(`Pref key '${styleText.blue(key)}' in ${styleText.gray(path)} is prefixed to ${styleText.blue(prefixed)}.`);
          content = content.replace(matched, `preference="${prefixed}"`);
        }
      }
      await outputFile(path, content, "utf-8");
    }));
  }
}
function renderPluginPrefsDts(prefs) {
  return `// Generated by zotero-plugin-scaffold
/* prettier-ignore */
/* eslint-disable */
// @ts-nocheck

// prettier-ignore
declare namespace _ZoteroTypes {
  interface Prefs {
    PluginPrefsMap: {
      ${Object.entries(prefs).map(([key, value]) => {
    return `"${key}": ${typeof value};`;
  }).join("\n      ")}
    };
  }
}
`;
}

class Build extends Base {
  buildTime;
  isPreRelease;
  constructor(ctx) {
    super(ctx);
    process.env.NODE_ENV ??= "production";
    this.buildTime = "";
    this.isPreRelease = this.ctx.version.includes("-");
  }
  /**
   * Default build runner
   */
  async run() {
    const { dist, version } = this.ctx;
    const t = /* @__PURE__ */ new Date();
    this.buildTime = dateFormat("YYYY-mm-dd HH:MM:SS", t);
    this.logger.info(
      `Building version ${styleText.blue(version)} to ${styleText.blue(dist)} at ${styleText.blue(this.buildTime)} in ${styleText.blue(process.env.NODE_ENV)} mode.`
    );
    await this.ctx.hooks.callHook("build:init", this.ctx);
    await emptyDir(dist);
    await this.ctx.hooks.callHook("build:mkdir", this.ctx);
    this.logger.tip("Preparing static assets", { space: 1 });
    await this.makeAssets();
    await this.ctx.hooks.callHook("build:copyAssets", this.ctx);
    this.logger.debug("Preparing manifest", { space: 2 });
    await this.makeManifest();
    await this.ctx.hooks.callHook("build:makeManifest", this.ctx);
    this.logger.debug("Preparing locale files", { space: 2 });
    await this.prepareLocaleFiles();
    await this.ctx.hooks.callHook("build:fluent", this.ctx);
    await this.preparePrefs();
    this.logger.tip("Bundling scripts", { space: 1 });
    await this.esbuild();
    await this.ctx.hooks.callHook("build:bundle", this.ctx);
    if (process.env.NODE_ENV === "production") {
      this.logger.tip("Packing plugin", { space: 1 });
      await this.pack();
      await this.ctx.hooks.callHook("build:pack", this.ctx);
      await this.makeUpdateJson();
      await this.ctx.hooks.callHook("build:makeUpdateJSON", this.ctx);
    }
    await this.ctx.hooks.callHook("build:done", this.ctx);
    this.logger.success(
      `Build finished in ${((/* @__PURE__ */ new Date()).getTime() - t.getTime()) / 1e3} s.`
    );
  }
  /**
   * Copys files in `Config.build.assets` to `Config.dist`
   */
  async makeAssets() {
    const { source, dist, build } = this.ctx;
    const { assets, define } = build;
    const paths = await glob(assets, { ignore: ["node_modules", ".git", dist] });
    const newPaths = paths.map((p) => `${dist}/addon/${p.replace(new RegExp(toArray(source).join("|")), "")}`);
    await Promise.all(paths.map(async (file, i) => {
      await copy(file, newPaths[i]);
      this.logger.debug(`Copy ${file} to ${newPaths[i]}`);
    }));
    const replaceMap = new Map(
      Object.keys(define).map((key) => [
        new RegExp(`__${key}__`, "g"),
        define[key]
      ])
    );
    this.logger.debug(`replace map: ${replaceMap}`);
    await replaceInFile({
      files: newPaths,
      from: Array.from(replaceMap.keys()),
      to: Array.from(replaceMap.values()),
      isGlob: false
    });
  }
  /**
   * Override user's manifest
   *
   */
  async makeManifest() {
    if (!this.ctx.build.makeManifest.enable)
      return;
    const { name, id, updateURL, dist, version } = this.ctx;
    const userData = await readJSON(
      `${dist}/addon/manifest.json`
    );
    const template = {
      ...userData,
      ...!userData.name && name && { name },
      ...version && { version },
      manifest_version: 2,
      applications: {
        zotero: {
          id,
          update_url: updateURL
        }
      }
    };
    const data = toMerged(userData, template);
    this.logger.debug(`manifest: ${JSON.stringify(data, null, 2)}`);
    outputJSON(`${dist}/addon/manifest.json`, data, { spaces: 2 });
  }
  async prepareLocaleFiles() {
    const { dist, namespace, build } = this.ctx;
    await buildLocale(dist, namespace, build.fluent);
  }
  async preparePrefs() {
    const { dist } = this.ctx;
    await buildPrefs(dist, this.ctx.build.prefs);
  }
  esbuild() {
    const { dist, build: { esbuildOptions } } = this.ctx;
    const distAbsolute = resolve(dist);
    if (esbuildOptions.length === 0)
      return;
    esbuildOptions.map((option, i) => {
      if (option.outfile && !resolve(option.outfile).startsWith(distAbsolute)) {
        this.logger.debug(`'outfile' of esbuildOptions[${i}] is not in dist folder, it will be overwritten.`);
        option.outfile = `${dist}/${option.outfile}`;
      }
      if (option.outdir && !resolve(option.outdir).startsWith(distAbsolute)) {
        this.logger.debug(`'outdir' of esbuildOptions[${i}] is not in dist folder, it will be overwritten.`);
        option.outdir = `${dist}/${option.outdir}`;
      }
      return option;
    });
    return Promise.all(
      esbuildOptions.map(
        (esbuildOption) => build(esbuildOption)
      )
    );
  }
  async makeUpdateJson() {
    const { dist, xpiName, id, version, xpiDownloadLink, build } = this.ctx;
    const manifest = await readJSON(
      `${dist}/addon/manifest.json`
    );
    const min = manifest.applications?.zotero?.strict_min_version;
    const max = manifest.applications?.zotero?.strict_max_version;
    const updateHash = await generateHash(`${dist}/${xpiName}.xpi`, "sha512");
    const data = {
      addons: {
        [id]: {
          updates: [
            ...build.makeUpdateJson.updates,
            {
              version,
              update_link: xpiDownloadLink,
              ...build.makeUpdateJson.hash && {
                update_hash: updateHash
              },
              applications: {
                zotero: {
                  ...min && { strict_min_version: min },
                  ...max && { strict_max_version: max }
                }
              }
            }
          ]
        }
      }
    };
    await writeJson(`${dist}/update-beta.json`, data, { spaces: 2 });
    if (!this.isPreRelease)
      await writeJson(`${dist}/update.json`, data, { spaces: 2 });
    this.logger.debug(
      `Prepare Update.json for ${this.isPreRelease ? "\x1B[31m Prerelease \x1B[0m" : "\x1B[32m Release \x1B[0m"}`
    );
  }
  async pack() {
    const { dist, xpiName } = this.ctx;
    const zip = new AdmZip();
    zip.addLocalFolder(`${dist}/addon`);
    zip.writeZip(`${dist}/${xpiName}.xpi`);
  }
}

class Bump extends Base {
  constructor(ctx) {
    super(ctx);
  }
  async run() {
    const bumppConfig = {
      ...this.ctx.release.bumpp,
      push: true,
      progress: this.bumppProgress
    };
    const { version } = this.ctx;
    if (bumppConfig.release === version) {
      this.logger.debug("Commit, tag, and push are disabled because new version = old version.");
      bumppConfig.commit = false;
      bumppConfig.tag = false;
      bumppConfig.push = false;
    }
    const result = await versionBump(bumppConfig);
    this.ctx.version = result.newVersion;
    this.ctx.release.bumpp.tag = result.tag || this.ctx.release.bumpp.tag.toString().replace("%s", result.newVersion);
    this.ctx.release.bumpp.commit = result.commit || this.ctx.release.bumpp.commit.toString().replace("%s", result.newVersion);
    this.logger.debug(`The release context after bump: ", ${this.ctx.release}`);
  }
  /**
   * bumpp 显示进度的回调
   *
   * @see https://github.com/antfu/bumpp/blob/main/src/cli/index.ts
   */
  get bumppProgress() {
    return ({
      event,
      script,
      updatedFiles,
      skippedFiles,
      newVersion
    }) => {
      switch (event) {
        case ProgressEvent.FileUpdated:
          this.logger.success(`Updated ${updatedFiles.pop()} to ${newVersion}`);
          break;
        case ProgressEvent.FileSkipped:
          this.logger.info(`${skippedFiles.pop()} did not need to be updated`);
          break;
        case ProgressEvent.GitCommit:
          this.logger.success("Git commit");
          break;
        case ProgressEvent.GitTag:
          this.logger.success("Git tag");
          break;
        case ProgressEvent.GitPush:
          this.logger.success("Git push");
          break;
        case ProgressEvent.NpmScript:
          this.logger.success(`Npm run ${script}`);
          break;
      }
    };
  }
}

const mimeTypes = {
  json: "application/json",
  xpi: "application/x-xpinstall"
};
function getMimeTypeByFileName(filename) {
  const ext = extname(filename);
  for (const type in mimeTypes) {
    if (ext === `.${type}`)
      return mimeTypes[type];
  }
  return void 0;
}

class ReleaseBase extends Base {
  isCI;
  constructor(ctx) {
    super(ctx);
    this.isCI = isCI;
  }
  checkFiles() {
    const { dist } = this.ctx;
    if (globSync(`${dist}/*.xpi`).length === 0) {
      throw new Error("No xpi file found, are you sure you have run build?");
    }
  }
}

class GitHub extends ReleaseBase {
  client;
  constructor(ctx) {
    super(ctx);
    this.client = this.getClient();
  }
  async run() {
    this.checkFiles();
    this.logger.info("Uploading XPI to GitHub...");
    await this.uploadXPI();
    this.logger.info("Refreshing update manifest...");
    await this.refreshUpdateManifest();
    return this.ctx;
  }
  /**
   * Create new release and upload XPI to asset
   */
  async uploadXPI() {
    const { version, dist, xpiName } = this.ctx;
    const release = await this.createRelease({
      ...this.remote,
      tag_name: this.ctx.release.bumpp.tag.toString().replaceAll("%s", version),
      name: `Release v${version}`,
      body: await this.getChangelog(),
      prerelease: version.includes("-"),
      make_latest: "true"
    });
    if (!release)
      throw new Error("Create release failed!");
    this.logger.debug("Uploading xpi asset...");
    await this.uploadAsset(release.id, join(dist, `${xpiName}.xpi`));
  }
  async getReleaseByTag(tag) {
    return await this.client.rest.repos.getReleaseByTag({
      ...this.remote,
      tag
    }).catch((e) => {
      this.logger.debug(`Release with tag ${tag} not found. ${e}`);
      return void 0;
    }).then((res) => {
      if (res && res.status === 200) {
        this.logger.debug(`Found release with tag "${tag}", id=${res.data.id}.`);
        return res.data;
      }
    });
  }
  async createRelease(options) {
    this.logger.debug("Creating release...");
    this.logger.debug(options);
    return await this.client.rest.repos.createRelease(options).catch((e) => {
      this.logger.error(e);
      throw new Error("Create release failed.");
    }).then((res) => {
      if (res.status === 201) {
        this.logger.debug(`Create release "${res.data.tag_name}" success, id: ${res.data.id}.`);
        return res.data;
      }
    });
  }
  async uploadAsset(releaseID, asset) {
    this.logger.debug(`Uploading ${asset} to release ${releaseID}`);
    return await this.client.rest.repos.uploadReleaseAsset({
      ...this.remote,
      release_id: releaseID,
      data: await readFile(asset),
      headers: {
        "content-type": getMimeTypeByFileName(asset) || "application/octet-stream",
        "content-length": (await stat(asset)).size
      },
      name: basename(asset)
    }).then((res) => {
      this.logger.debug(`Upload "${res.data.name}" success, assetId: ${res.data.id}`);
      return res.data;
    });
  }
  async refreshUpdateManifest() {
    const updater = this.ctx.release.github.updater;
    if (!updater) {
      this.logger.debug(`Skip refresh update.json because release.github.updater = false`);
      return;
    }
    const { dist, version } = this.ctx;
    const assets = (await glob(`${dist}/update*.json`)).map((p) => basename(p));
    const release = await this.getReleaseByTag(updater) ?? await this.createRelease({
      ...this.remote,
      tag_name: updater,
      prerelease: true,
      make_latest: "false"
    });
    if (!release)
      throw new Error("Get or create 'release' failed.");
    const existAssets = await this.client.rest.repos.listReleaseAssets({
      ...this.remote,
      release_id: release.id
    }).then((res) => {
      return res.data.filter((asset) => assets.includes(asset.name));
    });
    if (existAssets) {
      for (const existAsset of existAssets) {
        if (assets.includes(existAsset.name)) {
          this.logger.debug(`Delete existed asset ${existAsset.name} in release ${updater}`);
          await this.client.rest.repos.deleteReleaseAsset({
            ...this.remote,
            asset_id: existAsset.id
          });
        }
      }
    }
    for (const asset of assets) {
      await this.uploadAsset(release.id, join(dist, asset));
    }
    await this.client.rest.repos.updateRelease({
      ...this.remote,
      release_id: release.id,
      name: "Release Manifest",
      body: `This release is used to host \`update.json\`, please do not delete or modify it! 
 Updated in UTC ${(/* @__PURE__ */ new Date()).toISOString()} for version ${version}`,
      prerelease: true,
      make_latest: "false"
    });
  }
  async getChangelog() {
    const { release } = this.ctx;
    const { github } = release;
    const { releaseNote } = github;
    return releaseNote(this.ctx);
  }
  getClient() {
    if (!process.env.GITHUB_TOKEN)
      throw new Error("No GITHUB_TOKEN.");
    const client = new Octokit({
      auth: process.env.GITHUB_TOKEN,
      userAgent: "zotero-plugin-scaffold"
    });
    return client;
  }
  get remote() {
    const [owner, repo] = this.ctx.release.github.repository.split("/");
    return {
      owner,
      repo
    };
  }
}

class Release extends Base {
  constructor(ctx) {
    super(ctx);
  }
  /**
   * Runs release
   *
   * if is not CI，bump version, git add (package.json), git commit, git tag, git push;
   * if is CI, do not bump version, do not run git, create release (tag is `v${version}`) and upload xpi,
   *    then, create or update release (tag is "release"), update `update.json`.
   */
  async run() {
    const t = /* @__PURE__ */ new Date();
    const { release, version } = this.ctx;
    if (release.bumpp.release === "prompt" && isCI) {
      this.logger.warn("Config `release.bumpp.release == 'prompt'` will do nothing because in CI enviroment.");
      this.ctx.release.bumpp.release = version;
    }
    if (release.bumpp.confirm && isCI) {
      this.logger.warn("Config `release.bumpp.confirm` will do nothing because in CI enviroment.");
      this.ctx.release.bumpp.confirm = false;
    }
    const isBumpNeeded = this.ctx.release.bumpp.release !== version;
    const isGitHubEnabled = this.isEnabled(release.github.enable);
    const isPublishNeeded = isGitHubEnabled;
    if (isPublishNeeded && !release.bumpp.execute) {
      this.logger.warn(`The current release needs to run the build after bumping the version number, please configure the build script in 'config.release.bumpp.execute'${isBumpNeeded ? "" : " or run build before run release"}.`);
      this.ctx.release.bumpp.execute ||= "npm run build";
    }
    this.logger.debug(`Release config: ", ${this.ctx.release}`);
    await this.ctx.hooks.callHook("release:init", this.ctx);
    await new Bump(this.ctx).run();
    await this.ctx.hooks.callHook("release:push", this.ctx);
    this.ctx.release.changelog = await this.getChangelog();
    if (isGitHubEnabled) {
      await new GitHub(this.ctx).run();
    }
    await this.ctx.hooks.callHook("release:done", this.ctx);
    this.logger.success(
      `Done in ${((/* @__PURE__ */ new Date()).getTime() - t.getTime()) / 1e3} s.`
    );
  }
  async getConventionalChangelog() {
    const { version } = this.ctx;
    const { default: conventionalChangelog } = await import('conventional-changelog');
    return new Promise((resolve, reject) => {
      let changelog = "";
      conventionalChangelog({ releaseCount: 2, preset: "angular" }, { version }).on("data", (chunk) => {
        changelog += chunk.toString();
      }).on("end", () => {
        changelog = changelog.split("\n").filter((line) => !line.match(/^## .*/) && !line.match(/^# .*/)).join("\n").trim();
        resolve(changelog);
      }).on("error", (err) => {
        reject(err);
      });
    });
  }
  getGitLog() {
    const currentTag = this.ctx.release.bumpp.tag;
    const tags = execSync("git tag -l --sort=v:refname").toString().trim().split("\n");
    const currentTagIndex = tags.indexOf(currentTag);
    if (currentTagIndex === -1)
      throw new Error(`Tag "${currentTag}" not found.`);
    let previousTagIndex = currentTagIndex - 1;
    let previousTag;
    if (currentTagIndex === 0) {
      previousTag = false;
    } else if (currentTag.includes("-")) {
      previousTag = tags[previousTagIndex];
    } else {
      while (previousTagIndex >= 0 && tags[previousTagIndex].includes("-")) {
        previousTagIndex--;
      }
      if (previousTagIndex < 0)
        previousTag = false;
      else
        previousTag = tags[previousTagIndex];
    }
    if (previousTag)
      return execSync(`git log --pretty=format:"* %s (%h)" ${previousTag}..${currentTag}`).toString().trim();
    else
      return execSync(`git log --pretty=format:"* %s (%h)" ${currentTag}`).toString().trim();
  }
  getFilteredChangelog(rawLog) {
    const commitMessage = this.ctx.release.bumpp.commit;
    const filterRegex = new RegExp(escapeRegExp(commitMessage));
    const filteredLog = rawLog.split("\n").filter((line) => !filterRegex.test(line)).join("\n");
    if (filteredLog.replaceAll(" ", "") === "")
      return "_No significant changes._";
    return filteredLog;
  }
  async getChangelog() {
    let changelog;
    const changelogConfig = this.ctx.release.changelog;
    if (typeof changelogConfig == "function") {
      changelog = changelogConfig(this.ctx);
    } else if (!!changelogConfig && changelogConfig === "conventional-changelog") {
      const rawLog = await this.getConventionalChangelog();
      changelog = this.getFilteredChangelog(rawLog);
    } else if (!!changelogConfig && typeof changelogConfig == "string") {
      changelog = execSync(changelogConfig).toString().trim();
    } else {
      const rawLog = this.getGitLog();
      changelog = this.getFilteredChangelog(rawLog);
    }
    this.logger.debug(`Got changelog:
${changelog}
`);
    return changelog;
  }
  isEnabled(enable) {
    if (enable === "always")
      return true;
    if (enable === "ci" && isCI)
      return true;
    if (enable === "local" && !isCI)
      return true;
    return false;
  }
  get resolvedCommitMessage() {
    return this.ctx.release.bumpp.commit.replace("%s", this.ctx.version);
  }
}

function watch(source, event) {
  const watcher = chokidar.watch(source, {
    ignored: /(^|[/\\])\../,
    // ignore dotfiles
    persistent: true
  });
  const onChangeDebounced = safeDebounce(event.onChange);
  const onAddDebounced = safeDebounce(event.onAdd);
  watcher.on("ready", async () => {
    if (event.onReady)
      await event.onReady();
    logger.clear();
    logger.ready("Server Ready!");
  }).on("change", async (path) => {
    logger.clear();
    logger.info(`${path} changed`);
    await onChangeDebounced(path);
  }).on("add", async (path, stats) => {
    if (event.onAdd)
      await onAddDebounced(path, stats);
  }).on("error", async (err) => {
    if (event.onError)
      await event.onError(err);
    logger.fail("Server start failed!");
    logger.error(err);
  });
}
function safeDebounce(fn) {
  if (!fn)
    return () => {
    };
  return debounce(async (...args) => {
    try {
      await fn(...args);
    } catch (error) {
      logger.error(error);
    }
  }, 500);
}

function isRunning(query) {
  let cmd = "";
  switch (process.platform) {
    case "win32":
      cmd = `tasklist`;
      break;
    case "darwin":
      cmd = `ps -ax | grep ${query}`;
      break;
    case "linux":
      cmd = `ps -A`;
      break;
  }
  try {
    const stdout = execSync(cmd, { encoding: "utf8" });
    return stdout.toLowerCase().includes(query.toLowerCase());
  } catch {
    return false;
  }
}
const ExitSignals = ["exit", "SIGABRT", "SIGALRM", "SIGHUP", "SIGINT", "SIGTERM"];

const prefsCommon = {
  // Allow debug output via dump to be printed to the system console
  "browser.dom.window.dump.enabled": true,
  // From:
  // https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/internals/preferences.html#data-choices-notification
  // This is the data submission master kill switch. If disabled, no policy is shown or upload takes place, ever.
  "datareporting.policy.dataSubmissionEnabled": false,
  // Allow remote connections to the debugger.
  "devtools.debugger.remote-enabled": true,
  // Disable the prompt for allowing connections.
  "devtools.debugger.prompt-connection": false,
  // Allow extensions to log messages on browser's console.
  "devtools.browserconsole.contentMessages": true,
  // Turn off platform logging because it is a lot of info.
  "extensions.logging.enabled": false,
  // Disable extension updates and notifications.
  "extensions.checkCompatibility.nightly": false,
  "extensions.update.enabled": false,
  "extensions.update.notifyUser": false,
  // From:
  // http://hg.mozilla.org/mozilla-central/file/1dd81c324ac7/build/automation.py.in//l372
  // Only load extensions from the application and user profile.
  // AddonManager.SCOPE_PROFILE + AddonManager.SCOPE_APPLICATION
  "extensions.enabledScopes": 5,
  // Disable metadata caching for installed add-ons by default.
  "extensions.getAddons.cache.enabled": false,
  // Disable installing any distribution add-ons.
  "extensions.installDistroAddons": false,
  // Allow installing extensions dropped into the profile folder.
  "extensions.autoDisableScopes": 10,
  // Disable app update.
  "app.update.enabled": false,
  // Allow unsigned add-ons.
  "xpinstall.signatures.required": false,
  // browser.link.open_newwindow is changed from 3 to 2 in:
  // https://github.com/saadtazi/firefox-profile-js/blob/cafc793d940a779d280103ae17d02a92de862efc/lib/firefox_profile.js#L32
  // Restore original value to avoid https://github.com/mozilla/web-ext/issues/1592
  "browser.link.open_newwindow": 3
};
const prefsFirefox = {
  "browser.startup.homepage": "about:blank",
  "startup.homepage_welcome_url": "about:blank",
  "startup.homepage_welcome_url.additional": "",
  "devtools.errorconsole.enabled": true,
  "devtools.chrome.enabled": true,
  // From:
  // http://hg.mozilla.org/mozilla-central/file/1dd81c324ac7/build/automation.py.in//l388
  // Make url-classifier updates so rare that they won't affect tests.
  "urlclassifier.updateinterval": 172800,
  // Point the url-classifier to a nonexistent local URL for fast failures.
  "browser.safebrowsing.provider.0.gethashURL": "http://localhost/safebrowsing-dummy/gethash",
  "browser.safebrowsing.provider.0.keyURL": "http://localhost/safebrowsing-dummy/newkey",
  "browser.safebrowsing.provider.0.updateURL": "http://localhost/safebrowsing-dummy/update",
  // Disable self repair/SHIELD
  "browser.selfsupport.url": "https://localhost/selfrepair",
  // Disable Reader Mode UI tour
  "browser.reader.detectedFirstArticle": true,
  // Set the policy firstURL to an empty string to prevent
  // the privacy info page to be opened on every "web-ext run".
  // (See #1114 for rationale)
  "datareporting.policy.firstRunURL": ""
};
const prefsZotero = {
  "extensions.experiments.enabled": true,
  "extensions.autoDisableScopes": 0,
  // Enable remote-debugging
  "devtools.debugger.remote-enabled": true,
  "devtools.debugger.remote-websocket": true,
  "devtools.debugger.prompt-connection": false,
  // Inherit the default test settings from Zotero
  // https://github.com/zotero/zotero/blob/8a06edab49f07b84d07056ccd594c87920ef2c5e/test/runtests.sh#L127-L147
  // "app.update.enabled": false,
  // "extensions.zotero.sync.server.compressData": false,
  // "extensions.zotero.automaticScraperUpdates": false,
  // "extensions.zotero.debug.log": 5,
  // "extensions.zotero.debug.level": 5,
  // "extensions.zotero.debug.time": 5,
  "extensions.zotero.firstRun.skipFirefoxProfileAccessCheck": true,
  "extensions.zotero.firstRunGuidance": false,
  "extensions.zotero.firstRun2": false,
  "extensions.zotero.reportTranslationFailure": false,
  "extensions.zotero.httpServer.enabled": true,
  "extensions.zotero.httpServer.port": 23124,
  // ascii "ZT"
  "extensions.zotero.httpServer.localAPI.enabled": true,
  // "extensions.zotero.backup.numBackups": 0,
  // "extensions.zotero.sync.autoSync": false,
  "extensions.zoteroMacWordIntegration.installed": true,
  "extensions.zoteroMacWordIntegration.skipInstallation": true,
  "extensions.zoteroWinWordIntegration.skipInstallation": true,
  "extensions.zoteroOpenOfficeIntegration.skipInstallation": true
};
const prefs = {
  ...prefsCommon,
  ...prefsFirefox,
  ...prefsZotero
};

const UNSOLICITED_EVENTS = /* @__PURE__ */ new Set([
  "tabNavigated",
  "styleApplied",
  "propertyChange",
  "networkEventUpdate",
  "networkEvent",
  "propertyChange",
  "newMutations",
  "frameUpdate",
  "tabListChanged"
]);
function parseMessage(data) {
  const dataString = data.toString();
  const separatorIndex = dataString.indexOf(":");
  if (separatorIndex < 1) {
    return { remainingData: data };
  }
  const messageLength = Number.parseInt(dataString.substring(0, separatorIndex), 10);
  if (Number.isNaN(messageLength)) {
    return {
      remainingData: data,
      error: new Error("Error parsing RDP message length"),
      fatal: true
    };
  }
  if (data.length - (separatorIndex + 1) < messageLength) {
    return { remainingData: data };
  }
  const messageContent = data.slice(
    separatorIndex + 1,
    separatorIndex + 1 + messageLength
  );
  const remainingData = data.slice(separatorIndex + 1 + messageLength);
  try {
    const parsedMessage = JSON.parse(messageContent.toString());
    return { remainingData, parsedMessage };
  } catch (error) {
    return { remainingData, error, fatal: false };
  }
}
class MessagingClient extends EventEmitter {
  incomingData = Buffer.alloc(0);
  pendingRequests = [];
  activeRequests = /* @__PURE__ */ new Map();
  connection;
  constructor() {
    super();
  }
  async connect(port) {
    await new Promise((resolve, reject) => {
      const d = domain.create();
      d.once("error", reject);
      d.run(() => {
        const connectionOptions = { port, host: "127.0.0.1" };
        const conn = net.createConnection(connectionOptions, () => {
          resolve();
        });
        this.connection = conn;
        conn.on("data", this.onData.bind(this));
        conn.on("error", reject);
        conn.on("end", this.onEnd.bind(this));
        conn.on("timeout", this.onTimeout.bind(this));
        this.expectReply("root", { resolve, reject });
      });
    });
  }
  disconnect() {
    if (!this.connection)
      return;
    this.connection.removeAllListeners();
    this.connection.end();
    this.rejectAllRequests(
      new Error("RDP connection closed")
    );
  }
  rejectAllRequests(error) {
    this.activeRequests.forEach((deferred) => {
      deferred.reject(error);
    });
    this.activeRequests.clear();
    this.pendingRequests.forEach(({ deferred }) => {
      deferred.reject(error);
    });
    this.pendingRequests = [];
  }
  async request(requestProps) {
    const request = typeof requestProps === "string" ? { to: "root", type: requestProps } : requestProps;
    if (!request.to) {
      throw new Error(`Unexpected RDP request without target actor: ${request.type}`);
    }
    return await new Promise((resolve, reject) => {
      const deferred = { resolve, reject };
      this.pendingRequests.push({ request, deferred });
      this.flushPendingRequests();
    });
  }
  flushPendingRequests() {
    this.pendingRequests = this.pendingRequests.filter(
      ({ request, deferred }) => {
        if (this.activeRequests.has(request.to))
          return true;
        if (!this.connection) {
          throw new Error("RDP connection closed");
        }
        try {
          const messageString = `${Buffer.from(JSON.stringify(request)).length}:${JSON.stringify(request)}`;
          this.connection.write(messageString);
          this.expectReply(request.to, deferred);
        } catch (err) {
          deferred.reject(err);
        }
        return false;
      }
    );
  }
  expectReply(targetActor, deferred) {
    if (this.activeRequests.has(targetActor)) {
      throw new Error(`${targetActor} does already have an active request`);
    }
    this.activeRequests.set(targetActor, deferred);
  }
  onData(data) {
    this.incomingData = Buffer.concat([this.incomingData, data]);
    while (this.readMessage()) ;
  }
  readMessage() {
    const { remainingData, parsedMessage, error, fatal } = parseMessage(
      this.incomingData
    );
    this.incomingData = remainingData;
    if (error) {
      this.emit(
        "error",
        new Error(`Error parsing RDP packet: ${String(error)}`)
      );
      if (fatal)
        this.disconnect();
      return !fatal;
    }
    if (!parsedMessage)
      return false;
    this.handleMessage(parsedMessage);
    return true;
  }
  handleMessage(message) {
    if (!message.from) {
      if (message.error) {
        this.emit("rdp-error", message);
        return;
      }
      this.emit(
        "error",
        new Error(`Received an RDP message without a sender actor: ${JSON.stringify(message)}`)
      );
      return;
    }
    if (message.type && UNSOLICITED_EVENTS.has(message.type)) {
      this.emit("unsolicited-event", message);
      return;
    }
    if (this.activeRequests.has(message.from)) {
      const deferred = this.activeRequests.get(message.from);
      this.activeRequests.delete(message.from);
      if (message.error) {
        deferred?.reject(message);
      } else {
        deferred?.resolve(message);
      }
      this.flushPendingRequests();
      return;
    }
    if (message.type === "addonListChanged")
      return;
    this.emit(
      "error",
      new Error(`Received unexpected message:
 ${JSON.stringify(message)}`)
    );
  }
  onError(error) {
    this.emit("error", error);
  }
  onEnd() {
    this.emit("end");
  }
  onTimeout() {
    this.emit("timeout");
  }
}

const MAX_RETRIES = 150;
const RETRY_INTERVAL = 1e3;
function requestErrorToMessage(err) {
  if (err instanceof Error) {
    return String(err);
  }
  return `${err.error}: ${err.message}`;
}
function isErrorWithCode(codeWanted, error) {
  if (Array.isArray(codeWanted) && codeWanted.includes(error.code)) {
    return true;
  } else if (error.code === codeWanted) {
    return true;
  }
  return false;
}
class RemoteFirefox {
  client;
  checkedForAddonReloading;
  constructor() {
    this.client = new MessagingClient();
    this.checkedForAddonReloading = false;
    this.client.on("disconnect", () => {
      logger.debug('Received "disconnect" from Firefox client');
    });
    this.client.on("end", () => {
      logger.debug('Received "end" from Firefox client');
    });
    this.client.on("unsolicited-event", (info) => {
      logger.debug(`Received message from client: ${JSON.stringify(info)}`);
    });
    this.client.on("rdp-error", (rdpError) => {
      logger.debug(`Received error from client: ${JSON.stringify(rdpError)}`);
    });
    this.client.on("error", (error) => {
      logger.debug(`Received error from client: ${String(error)}`);
    });
  }
  async connect(port) {
    let lastError;
    for (const _ of Array.from({ length: MAX_RETRIES })) {
      try {
        this.client = new MessagingClient();
        await this.client.connect(port);
        return this.client;
      } catch (error) {
        logger.debug("Failed to connecte to RDP client, retry...");
        if (isErrorWithCode("ECONNREFUSED", error)) {
          await new Promise((resolve) => setTimeout(resolve, RETRY_INTERVAL));
          lastError = error;
        } else {
          throw error;
        }
      }
    }
    logger.error(`Unable to connect to Zotero. Too many retries.`);
    throw lastError;
  }
  disconnect() {
    this.client.disconnect();
  }
  async addonRequest(addon, request) {
    try {
      const response = await this.client.request({
        to: addon.actor,
        type: request
      });
      return response;
    } catch (err) {
      logger.debug(`Client responded to '${request}' request with error: ${err}`);
      const message = requestErrorToMessage(err);
      throw new Error(`Remote Firefox: addonRequest() error: ${message}`);
    }
  }
  async getAddonsActor() {
    try {
      const response = await this.client.request("getRoot");
      if (response.addonsActor == null) {
        return Promise.reject(
          new Error(
            "This version of Firefox does not provide an add-ons actor for remote installation."
          )
        );
      }
      return response.addonsActor;
    } catch (err) {
      logger.debug(`Falling back to listTabs because getRoot failed", ${err}`);
    }
    try {
      const response = await this.client.request("listTabs");
      if (response.addonsActor == null) {
        logger.debug(
          `listTabs returned a falsey addonsActor: ${JSON.stringify(response)}`
        );
        return Promise.reject(
          new Error(
            "This is an older version of Firefox that does not provide an add-ons actor for remote installation. Try Firefox 49 or higher."
          )
        );
      }
      return response.addonsActor;
    } catch (err) {
      logger.debug(`listTabs error: ${err}`);
      const message = requestErrorToMessage(err);
      throw new Error(`Remote Firefox: listTabs() error: ${message}`);
    }
  }
  async installTemporaryAddon(addonPath) {
    const addonsActor = await this.getAddonsActor();
    try {
      const response = await this.client.request({
        to: addonsActor,
        type: "installTemporaryAddon",
        addonPath
        // openDevTools: true,
      });
      logger.debug(`installTemporaryAddon: ${JSON.stringify(response)}`);
      logger.info(`Installed ${addonPath} as a temporary add-on`);
      return response;
    } catch (err) {
      const message = requestErrorToMessage(err);
      throw new Error(`installTemporaryAddon: Error: ${message}`);
    }
  }
  async getInstalledAddon(addonId) {
    try {
      const response = await this.client.request("listAddons");
      for (const addon of response.addons) {
        if (addon.id === addonId) {
          return addon;
        }
      }
      logger.debug(
        `Remote Firefox has these addons: ${response.addons.map((a) => a.id)}`
      );
      return Promise.reject(
        new Error(
          "The remote Firefox does not have your extension installed"
        )
      );
    } catch (err) {
      const message = requestErrorToMessage(err);
      throw new Error(`Remote Firefox: listAddons() error: ${message}`);
    }
  }
  async checkForAddonReloading(addon) {
    if (this.checkedForAddonReloading) {
      return addon;
    } else {
      const response = await this.addonRequest(addon, "requestTypes");
      if (!response.requestTypes.includes("reload")) {
        const supportedRequestTypes = JSON.stringify(response.requestTypes);
        logger.debug(`Remote Firefox only supports: ${supportedRequestTypes}`);
        throw new Error(
          "This Firefox version does not support add-on reloading. Re-run with --no-reload"
        );
      } else {
        this.checkedForAddonReloading = true;
        return addon;
      }
    }
  }
  async reloadAddon(addonId) {
    const addon = await this.getInstalledAddon(addonId);
    await this.checkForAddonReloading(addon);
    await this.addonRequest(addon, "reload");
    logger.success(`Last extension reload: ${(/* @__PURE__ */ new Date()).toTimeString()}`);
  }
}
function findFreeTcpPort() {
  return new Promise((resolve) => {
    const srv = net.createServer();
    srv.listen(0, "127.0.0.1", () => {
      const addr = srv.address();
      if (addr && typeof addr !== "string") {
        const freeTcpPort = addr.port;
        srv.close(() => resolve(freeTcpPort));
      }
    });
  });
}

const default_options = {
  binary: {
    // path: "",
    args: [],
    devtools: true
  },
  profile: {
    path: "./.scaffold/profile",
    dataDir: "",
    // keepChanges: true,
    createIfMissing: true,
    customPrefs: {}
  },
  plugins: {
    asProxy: false,
    list: []
  }
};
class ZoteroRunner {
  options;
  remoteFirefox = new RemoteFirefox();
  zotero;
  constructor(options) {
    this.options = toMerged(default_options, options);
    if (!options.binary.path)
      throw new Error("Binary path must be provided.");
    if (!options.profile.path && !options.profile.dataDir)
      this.options.profile.dataDir = "./.scaffold/data";
    logger.debug(this.options);
  }
  get default_profile_path() {
    return default_options.profile.path;
  }
  async run() {
    await this.setupProfile();
    await this.startZoteroInstance();
    if (!this.options.plugins.asProxy)
      await this.installTemporaryPlugins();
  }
  /**
   * Preparing the development environment
   *
   * When asProxy=true, generate a proxy file and replace prefs.
   *
   * @see https://www.zotero.org/support/dev/client_coding/plugin_development#setting_up_a_plugin_development_environment
   */
  async setupProfile() {
    const { path, createIfMissing } = this.options.profile;
    if (!await pathExists(path)) {
      if (createIfMissing)
        await this.createProfile(this.default_profile_path);
      else
        throw new Error("The 'profile.path' must be provided when 'createIfMissing' is false.");
    }
    const prefsPath = join(this.options.profile.path, "prefs.js");
    const prefsManager = new PrefsManager("user_pref");
    prefsManager.setPrefs(prefs);
    prefsManager.setPrefs(this.options.profile.customPrefs);
    if (await pathExists(prefsPath))
      await prefsManager.read(prefsPath);
    prefsManager.setPrefs({
      "extensions.lastAppBuildId": null,
      "extensions.lastAppVersion": null
    });
    await prefsManager.write(prefsPath);
    if (this.options.plugins.asProxy) {
      await this.installProxyPlugins();
    }
  }
  async createProfile(path) {
    logger.debug(`Creating profile at ${this.default_profile_path}...`);
    await ensureDir(path);
  }
  // private async copyProfile(from: string, to = this.default_profile_path) {
  //   logger.debug(`Copying profile from '${this.options.profile.path}' to ${this.default_profile_path}...`);
  //   await copy(from, to);
  //   this.options.profile.path = to;
  // }
  async startZoteroInstance() {
    let args = ["--purgecaches", "no-remote"];
    if (this.options.profile.path) {
      args.push("-profile", resolve(this.options.profile.path));
    }
    if (this.options.profile.dataDir) {
      args.push("--dataDir", resolve(this.options.profile.dataDir));
    }
    if (this.options.binary.devtools) {
      args.push("--jsdebugger");
    }
    if (this.options.binary.args) {
      args = [...args, ...this.options.binary.args];
    }
    const remotePort = await findFreeTcpPort();
    args.push("-start-debugger-server", String(remotePort));
    logger.debug(`Zotero start args: ${args}`);
    const env = {
      ...process.env,
      XPCOM_DEBUG_BREAK: "stack",
      NS_TRACE_MALLOC_DISABLE_STACKS: "1"
    };
    if (!await pathExists(this.options.binary.path))
      throw new Error("The Zotero binary not found.");
    this.zotero = spawn(this.options.binary.path, args, { env });
    logger.debug(`Zotero started, pid: ${this.zotero.pid}`);
    this.zotero.stdout?.on("data", (_data) => {
    });
    logger.debug("Connecting to the remote Firefox debugger...");
    await this.remoteFirefox.connect(remotePort);
    logger.debug(`Connected to the remote Firefox debugger on port: ${remotePort}`);
  }
  async installTemporaryPlugins() {
    for (const plugin of this.options.plugins.list) {
      const addonId = await this.remoteFirefox.installTemporaryAddon(resolve(plugin.sourceDir)).then((installResult) => {
        return installResult.addon.id;
      });
      if (!addonId) {
        throw new Error("Unexpected missing addonId in the installAsTemporaryAddon result");
      }
    }
  }
  async installProxyPlugin(id, sourceDir) {
    const addonProxyFilePath = join(this.options.profile.path, `extensions/${id}`);
    const buildPath = resolve(sourceDir);
    await outputFile(addonProxyFilePath, buildPath);
    logger.debug(
      [
        `Addon proxy file has been updated.`,
        `  File path: ${addonProxyFilePath}`,
        `  Addon path: ${buildPath}`
      ].join("\n")
    );
    const addonXpiFilePath = join(this.options.profile.path, `extensions/${id}.xpi`);
    if (await pathExists(addonXpiFilePath)) {
      await remove(addonXpiFilePath);
      logger.debug(`XPI file found, removed.`);
    }
    const addonInfoFilePath = join(this.options.profile.path, "extensions.json");
    if (await pathExists(addonInfoFilePath)) {
      const content = await readJSON(addonInfoFilePath);
      content.addons = content.addons.map((addon) => {
        if (addon.id === id && addon.active === false) {
          addon.active = true;
          addon.userDisabled = false;
          logger.debug(`Active plugin ${id} in extensions.json.`);
        }
        return addon;
      });
      await outputJSON(addonInfoFilePath, content);
    }
  }
  async installProxyPlugins() {
    for (const { id, sourceDir } of this.options.plugins.list) {
      await this.installProxyPlugin(id, sourceDir);
    }
  }
  async reloadTemporaryPluginById(id) {
    await this.remoteFirefox.reloadAddon(id);
  }
  async reloadTemporaryPluginBySourceDir(sourceDir) {
    const addonId = this.options.plugins.list.find((p) => p.sourceDir === sourceDir)?.id;
    if (!addonId) {
      return {
        sourceDir,
        reloadError: new Error(
          `Extension not reloadable: no addonId has been mapped to "${sourceDir}"`
        )
      };
    }
    try {
      await this.remoteFirefox.reloadAddon(addonId);
    } catch (error) {
      return {
        sourceDir,
        reloadError: error
      };
    }
    return { sourceDir, reloadError: void 0 };
  }
  async reloadAllTemporaryPlugins() {
    for (const { sourceDir } of this.options.plugins.list) {
      const res = await this.reloadTemporaryPluginBySourceDir(sourceDir);
      if (res.reloadError instanceof Error) {
        logger.error(res.reloadError);
      }
    }
  }
  async reloadProxyPluginByZToolkit(id, name, version) {
    const reloadScript = `
    (async () => {
    Services.obs.notifyObservers(null, "startupcache-invalidate", null);
    const { AddonManager } = ChromeUtils.import("resource://gre/modules/AddonManager.jsm");
    const addon = await AddonManager.getAddonByID("${id}");
    await addon.reload();
    const progressWindow = new Zotero.ProgressWindow({ closeOnClick: true });
    progressWindow.changeHeadline("${name} Hot Reload");
    progressWindow.progress = new progressWindow.ItemProgress(
        "chrome://zotero/skin/tick.png",
        "VERSION=${version}, BUILD=${(/* @__PURE__ */ new Date()).toLocaleString()}. By zotero-plugin-toolkit"
    );
    progressWindow.progress.setProgress(100);
    progressWindow.show();
    progressWindow.startCloseTimer(5000);
    })()`;
    const url = `zotero://ztoolkit-debug/?run=${encodeURIComponent(
      reloadScript
    )}`;
    const startZoteroCmd = `"${this.options.binary.path}" --purgecaches -profile "${this.options.profile.path}"`;
    const command = `${startZoteroCmd} -url "${url}"`;
    execSync(command);
  }
  // Do not use this method if possible,
  // as frequent execSync can cause Zotero to crash.
  async reloadAllProxyPlugins() {
    for (const { id } of this.options.plugins.list) {
      await this.reloadProxyPluginByZToolkit(id, id, id);
      await delay(2e3);
    }
  }
  async reloadAllPlugins() {
    if (this.options.plugins.asProxy)
      await this.reloadAllProxyPlugins();
    else
      await this.reloadAllTemporaryPlugins();
  }
  exit() {
    this.zotero?.kill();
    killZotero();
  }
}
function killZotero() {
  function kill() {
    try {
      if (process.env.ZOTERO_PLUGIN_KILL_COMMAND) {
        execSync(process.env.ZOTERO_PLUGIN_KILL_COMMAND);
      } else if (isWindows) {
        execSync("taskkill /f /im zotero.exe");
      } else if (isMacOS) {
        execSync("kill -9 $(ps -x | grep zotero)");
      } else if (isLinux) {
        execSync("pkill -9 zotero");
      } else {
        logger.error("No commands found for this operating system.");
      }
    } catch {
      logger.fail("Kill Zotero failed.");
    }
  }
  if (isRunning("zotero")) {
    kill();
  } else {
    logger.fail("No Zotero instance is currently running.");
  }
}

class Serve extends Base {
  builder;
  runner;
  _zoteroBinPath;
  constructor(ctx) {
    super(ctx);
    process.env.NODE_ENV ??= "development";
    this.builder = new Build(ctx);
  }
  async run() {
    process.on("SIGINT", this.exit);
    this.runner = new ZoteroRunner({
      binary: {
        path: this.zoteroBinPath,
        devtools: this.ctx.server.devtools,
        args: this.ctx.server.startArgs
      },
      profile: {
        path: this.profilePath,
        dataDir: this.dataDir,
        // keepChanges: this.ctx.server.keepProfileChanges,
        createIfMissing: this.ctx.server.createProfileIfMissing
      },
      plugins: {
        list: [{
          id: this.ctx.id,
          sourceDir: join(this.ctx.dist, "addon")
        }],
        asProxy: this.ctx.server.asProxy
      }
    });
    await this.ctx.hooks.callHook("serve:init", this.ctx);
    if (this.ctx.server.prebuild) {
      await this.builder.run();
      await this.ctx.hooks.callHook("serve:prebuild", this.ctx);
    }
    await this.runner.run();
    this.runner.zotero?.on("exit", this.onZoteroExit);
    this.runner.zotero?.on("close", this.onZoteroExit);
    await this.watch();
  }
  /**
   * watch source dir and build when file changed
   */
  async watch() {
    const { source } = this.ctx;
    watch(
      source,
      {
        onReady: async () => {
          await this.ctx.hooks.callHook("serve:ready", this.ctx);
        },
        onChange: async (path) => {
          await this.ctx.hooks.callHook("serve:onChanged", this.ctx, path);
          if (path.endsWith(".ts") || path.endsWith(".tsx")) {
            await this.builder.esbuild();
          } else {
            await this.builder.run();
          }
          await this.reload();
        }
      }
    );
  }
  async reload() {
    this.logger.tip("Reloading...");
    await this.runner?.reloadAllPlugins();
    await this.ctx.hooks.callHook("serve:onReloaded", this.ctx);
  }
  // Use arrow functions to keep `this`
  exit = () => {
    this.logger.info("Server shutdown by user request.");
    this.runner?.exit();
    this.ctx.hooks.callHook("serve:exit", this.ctx);
    process.exit();
  };
  onZoteroExit = (_code, _signal) => {
    this.logger.info(`Zotero terminated.`);
    process.exit();
  };
  get zoteroBinPath() {
    if (this._zoteroBinPath)
      return this._zoteroBinPath;
    this._zoteroBinPath = process.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH;
    if (!this._zoteroBinPath || !existsSync(this._zoteroBinPath))
      throw new Error("The Zotero binary not found.");
    return this._zoteroBinPath;
  }
  get profilePath() {
    return process.env.ZOTERO_PLUGIN_PROFILE_PATH;
  }
  get dataDir() {
    return process.env.ZOTERO_PLUGIN_DATA_DIR;
  }
}

const BASE_DIR = ".scaffold";
const CACHE_DIR = `${BASE_DIR}/cache`;
const TESTER_BASE_PATH = `${BASE_DIR}/test`;
const TESTER_PROFILE_DIR = `${TESTER_BASE_PATH}/profile`;
const TESTER_DATA_DIR = `${TESTER_BASE_PATH}/data`;
const TESTER_PLUGIN_DIR = `${TESTER_BASE_PATH}/resource`;
const TESTER_PLUGIN_REF = "zotero-plugin-scaffold-test-runner";
const TESTER_PLUGIN_ID = "scaffold-test@northword.cn";

async function prepareHeadless() {
  await installXvfb();
  await installDepsForUbuntu24();
  await installZoteroLinux();
  const xvfb = new Xvfb({
    timeout: 2e3
  });
  await xvfb.start();
}
function isPackageInstalled(packageName) {
  try {
    execSync(`dpkg-query -W ${packageName}`, { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}
function installPackage(packageName) {
  const debug = isDebug || logger.level <= LOG_LEVEL.DEBUG;
  try {
    logger.debug(`Installing ${packageName}...`);
    execSync(`sudo apt update && sudo apt install -y ${packageName}`, { stdio: debug ? "inherit" : "pipe" });
    logger.debug(`${packageName} installed successfully.`);
  } catch (error) {
    logger.fail(`Failed to install ${packageName}. ${error}`);
    throw error;
  }
}
function checkAndInstallDependencies(packages) {
  packages.forEach((pkg) => {
    if (isPackageInstalled(pkg)) {
      logger.debug(`${pkg} is already installed.`);
    } else {
      installPackage(pkg);
    }
  });
}
async function installXvfb() {
  if (!isLinux) {
    logger.error("Unsupported platform. Please install Xvfb manually.");
    return;
  }
  const osId = execSync("cat /etc/os-release | grep '^ID='").toString();
  if (!(osId.includes("ubuntu") || osId.includes("debian"))) {
    logger.error("Unsupported Linux distribution.");
    return;
  }
  checkAndInstallDependencies(["xvfb", "x11-xkb-utils", "xkb-data"]);
}
async function installDepsForUbuntu24() {
  const version = execSync("cat /etc/os-release | grep '^VERSION_ID='").toString();
  if (!version.includes("24")) {
    logger.error("Skip to install deps due to version not 24.04.");
    return;
  }
  checkAndInstallDependencies(["libasound2t64", "libdbus-glib-1-2"]);
}
async function installZoteroLinux() {
  if (process.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH) {
    logger.debug("Local Zotero found, skip to download.");
    return;
  }
  logger.debug("Installing Zotero...");
  execSync(`cd ${CACHE_DIR}`);
  execSync("wget -O zotero.tar.bz2 'https://www.zotero.org/download/client/dl?platform=linux-x86_64&channel=beta'", { stdio: "pipe" });
  execSync("tar -xvf zotero.tar.bz2", { stdio: "pipe" });
  process.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH = `${process.cwd()}/Zotero_linux-x86_64/zotero`;
  execSync("cd -");
}

class TestHttpReporter {
  _server;
  _port;
  passed = 0;
  failed = 0;
  constructor() {
  }
  async getPort() {
    this._port = await findFreeTcpPort();
  }
  get port() {
    return this._port;
  }
  async start() {
    if (!this._port)
      await this.getPort();
    this._server = http.createServer(this.handleRequest.bind(this));
    this._server.listen(this._port, () => {
      logger.debug(`Server is listening on http://localhost:${this.port}`);
    });
    return this;
  }
  handleRequest(req, res) {
    if (req.method === "GET" && req.url === "/") {
      res.writeHead(200, { "Content-Type": "text/plain" });
      res.end("Zotero Plugin Test Server is running");
    } else if (req.method === "POST" && req.url === "/update") {
      let body = "";
      req.on("data", (chunk) => {
        body += chunk;
      });
      req.on("end", async () => {
        try {
          const jsonData = JSON.parse(body);
          await this.onData(jsonData);
          res.writeHead(200, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ message: "Results received successfully" }));
        } catch (error) {
          logger.error(`Error parsing JSON:, ${error}`);
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(JSON.stringify({ error: "Invalid JSON" }));
        }
      });
    } else {
      res.writeHead(404, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ error: "Not Found" }));
    }
  }
  async onData(body) {
    const { type, data } = body;
    const logger_option = { space: data.indents - 1 };
    switch (type) {
      case "debug":
        logger.log(data);
        break;
      case "start":
        logger.newLine();
        break;
      case "suite":
        if (data.title)
          logger.tip(data.title, logger_option);
        break;
      case "pass":
        logger.success(`${data.title} ${styleText.gray(`${data.duration}ms`)}`, logger_option);
        break;
      case "fail": {
        let formatOutput = function(obj, indent) {
          const jsonStr = JSON.stringify(obj, null, 2);
          const lines = jsonStr.split("\n");
          const firstLine = lines[0];
          const restLines = lines.slice(1).map((line) => "  ".repeat(indent) + line).join("\n");
          return `${firstLine}
${restLines}`;
        };
        logger.fail(styleText.red(`${data.title}, ${data?.error?.message}`), logger_option);
        let expected = data.error.expected;
        let received = data.error.actual;
        if (expected && typeof expected === "object")
          expected = formatOutput(expected, data.indents + 1);
        if (received && typeof received === "object")
          received = formatOutput(received, data.indents + 1);
        logger.log(`Expected: ${styleText.green(String(expected))}`, { space: data.indents + 0.5 });
        logger.log(`Received: ${styleText.red(String(received))}`, { space: data.indents + 0.5 });
        break;
      }
      case "pending":
        logger.info(`${data.title} pending`, logger_option);
        break;
      case "suite end":
        break;
      case "end":
        this.passed = data.passed;
        this.failed = data.failed;
        logger.newLine();
        if (this.failed === 0)
          logger.success(`Test run completed - ${this.passed} passed`);
        else
          logger.fail(`Test run completed - ${this.passed} passed, ${this.failed} failed`);
        break;
      default:
        logger.log(data);
        break;
    }
  }
  stop() {
    this._server?.close();
  }
}

async function saveResource(url, path) {
  const res = await fetch(url);
  await outputFile(path, await res.text());
}

function generateBootstrap(options) {
  return `//Code generated by the zotero-plugin-scaffold tester
var chromeHandle;

function install(data, reason) {}

async function startup({ id, version, resourceURI, rootURI }, reason) {
  await Zotero.initializationPromise;
  const aomStartup = Components.classes[
    "@mozilla.org/addons/addon-manager-startup;1"
  ].getService(Components.interfaces.amIAddonManagerStartup);
  const manifestURI = Services.io.newURI(rootURI + "manifest.json");
  chromeHandle = aomStartup.registerChrome(manifestURI, [
    ["content", "${TESTER_PLUGIN_REF}", rootURI + "content/"],
  ]);

  launchTests().catch((error) => {
    Zotero.debug(error);
    Zotero.HTTP.request(
      "POST",
      "http://localhost:${options.port}/update",
      {
        body: JSON.stringify({
          type: "fail",
          data: {
            title: "Internal: Plugin awaiting timeout",
            stack: "",
            str: "Plugin awaiting timeout",
          },
        }),
      }
    );
  });
}

function onMainWindowLoad({ window: win }) {}

function onMainWindowUnload({ window: win }) {}

function shutdown({ id, version, resourceURI, rootURI }, reason) {
  if (reason === APP_SHUTDOWN) {
    return;
  }

  if (chromeHandle) {
    chromeHandle.destruct();
    chromeHandle = null;
  }
}

function uninstall(data, reason) {}

async function launchTests() {
  // Delay to allow plugin to fully load before opening the test page
  await Zotero.Promise.delay(${options.startupDelay || 1e3});

  const waitForPlugin = "${options.waitForPlugin}";

  if (waitForPlugin) {
    // Wait for a plugin to be installed
    await waitUtilAsync(() => {
      try {
        return !!eval(waitForPlugin)();
      } catch (error) {
        return false;
      }
    }).catch(() => {
      throw new Error("Plugin awaiting timeout");
    });
  }

  Services.ww.openWindow(
    null,
    "chrome://${TESTER_PLUGIN_REF}/content/index.xhtml",
    "Zotero Plugin Scaffold Test Runnner",
    "chrome,centerscreen,resizable=yes",
    {}
  );
}

function waitUtilAsync(condition, interval = 100, timeout = 1e4) {
  return new Promise((resolve, reject) => {
    const start = Date.now();
    const intervalId = setInterval(() => {
      if (condition()) {
        clearInterval(intervalId);
        resolve();
      } else if (Date.now() - start > timeout) {
        clearInterval(intervalId);
        reject();
      }
    }, interval);
  });
}
`;
}

function generateHtml(setupCode, testFiles) {
  const tests = testFiles.map((f) => `<script src="${f}"><\/script>`).join("\n    ");
  return `<!-- Generated by zotero-plugin-scaffold -->
<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head>
    <meta charset="UTF-8"></meta>
    <meta name="viewport" content="width=device-width, initial-scale=1.0"></meta>
    <title>Zotero Plugin Test</title>
    <style>
        html {
            min-width: 400px;
            min-height: 600px;
        }
        body {
            font-family: Arial, sans-serif;
        }
    </style>
</head>
<body>
    <div id="mocha"></div>

    <!-- Include Zotero Vars -->
    <script src="chrome://zotero/content/include.js"><\/script>

    <!-- Mocha and Chai Libraries -->
    <script src="mocha.js"><\/script>
    <script src="chai.js"><\/script>

    <!-- Setup Mocha -->
    <script>
      ${setupCode}
    <\/script>

    <!-- Unit tests -->
    ${tests}

    <!-- Run Mocha -->
    <script class="mocha-exec">
      mocha.run();
    <\/script>
</body>
</html>
`;
}

function generateManifest() {
  return {
    manifest_version: 2,
    name: "Zotero Plugin Scaffold Test Runner",
    version: "0.0.1",
    description: "Test suite for the Zotero plugin. This is a runtime-generated plugin only for testing purposes.",
    applications: {
      zotero: {
        id: TESTER_PLUGIN_ID,
        update_url: "https://example.com",
        // strict_min_version: "*.*.*",
        strict_max_version: "999.*.*"
      }
    }
  };
}

function generateMochaSetup(options) {
  return `// Generated by zotero-plugin-scaffold
mocha.setup({ 
  ui: "bdd",
  reporter: Reporter,
  timeout: ${options.timeout} || 10000,
  bail: ${options.abortOnFail}
});

window.expect = chai.expect;
window.assert = chai.assert;

async function send(data) {
  const req = await Zotero.HTTP.request(
    "POST",
    "http://localhost:${options.port}/update",
    {
      body: JSON.stringify(data),
    }
  );

  if (req.status !== 200) {
    dump("Error sending data to server" + req.responseText);
    return null;
  } else {
    const result = JSON.parse(req.responseText);
    return result;
  }
}

window.debug = function (data) {
  send({ type: "debug", data });
};

let indents = 0,
  passed = 0,
  failed = 0,
  aborted = false;

// Inherit the default test settings from Zotero
function Reporter(runner) {
  function indent() {
    return "  ".repeat(indents);
  }

  function dump(str) {
    // console.log(str)
    // const p = document.createElement("p")
    // p.textContent = str
    // document.querySelector("#mocha").append(p)
    document.querySelector("#mocha").innerText += str;
  }

  runner.on("start", async function () {
    console.log("start")
    await send({ type: "start", data: { indents } });
  });

  runner.on("suite", async function (suite) {
    console.log("suite", suite)
    indents++;
    const str = indent() + suite.title + "\\n";
    dump(str);
    await send({ type: "suite", data: { title: suite.title, root: suite.root, indents } });
  });

  runner.on("suite end", async function (suite) {
    console.log("suite end", suite)
    indents--;
    const str = indents === 1 ? "\\n" : "";
    dump(str);
    await send({ type: "suite end", data: { title: suite.title, root: suite.root, indents } });
  });

  runner.on("pending", async function (test) {
    console.log("pending", test)
    const str = indent() + "pending  -" + test.title + "\\n";
    dump(str);
    await send({ type: "pending", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, indents: indents + 1 } });
  });

  runner.on("pass", async function (test) {
    console.log("pass", test)
    passed++;
    let str = indent() + Mocha.reporters.Base.symbols.ok + " " + test.title;
    if ("fast" != test.speed) {
      str += " (" + Math.round(test.duration) + " ms)";
    }
    str += "\\n";
    dump(str);
    await send({ type: "pass", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, indents: indents + 1 } });

  });

  runner.on("fail", async function (test, error) {
    console.log("fail", test, error)

    // Make sure there's a blank line after all stack traces
    // err.stack = err.stack.replace(/\\s*$/, "\\n\\n");

    failed++;
    let indentStr = indent();
    const str =
      indentStr +
      // Dark red X for errors
      Mocha.reporters.Base.symbols.err +
      // Trigger bell if interactive
      (Zotero.automatedTest ? "" : "\\x07") +
      " " +
      test.title +
      "\\n" +
      indentStr +
      "  " +
      error.message +
      // " at\\n" +
      // err.stack.replace(/^/gm, indentStr + "    ").trim() +
      "\\n\\n";
    dump(str);

    await send({ type: "fail", data: { title: test.title, fulltest: test.fullTitle(), duration: test.duration, error, indents: indents + 1 } });

  });

  runner.on("end", async function () {
    console.log("end")
    const str =
      passed +
      "/" +
      (passed + failed) +
      " tests passed" +
      (aborted ? " -- aborting" : "") +
      "\\n";
    dump(str);

    await send({
      type: "end",
      data: { passed: passed, failed: failed, aborted: aborted, str, indents },
    });

    // Must exit on Zotero side, otherwise the exit code will not be 0 and CI will fail
    if (${options.exitOnFinish ? "true" : "false"}) {
      Zotero.Utilities.Internal.quit(0);
    }
  });
}
`;
}

class TestBundler {
  constructor(ctx, port) {
    this.ctx = ctx;
    this.port = port;
  }
  esbuildContext;
  async generate() {
    await this.generateTestResources();
    await this.createTestHtml();
  }
  async regenerate(changedFile) {
    const esbuildResult = await this.esbuildContext?.rebuild();
    const tests = findImpactedTests(changedFile, esbuildResult?.metafile);
    await this.createTestHtml(tests);
  }
  async generateTestResources() {
    const manifest = generateManifest();
    await outputJSON(`${TESTER_PLUGIN_DIR}/manifest.json`, manifest, { spaces: 2 });
    const bootstrap = generateBootstrap({
      port: this.port,
      startupDelay: this.ctx.test.startupDelay,
      waitForPlugin: this.ctx.test.waitForPlugin
    });
    await outputFile(`${TESTER_PLUGIN_DIR}/bootstrap.js`, bootstrap);
    await this.copyTestLibraries();
    await this.bundleTests();
  }
  async copyTestLibraries() {
    const pkgs = [
      {
        name: "mocha.js",
        local: "node_modules/mocha/mocha.js",
        remote: "https://cdn.jsdelivr.net/npm/mocha/mocha.js"
      },
      {
        name: "chai.js",
        // local: "node_modules/chai/chai.js",
        local: "",
        // chai packages install from npm do not support browser
        remote: "https://www.chaijs.com/chai.js"
      }
    ];
    await Promise.all(pkgs.map(async (pkg) => {
      const targetPath = `${TESTER_PLUGIN_DIR}/content/${pkg.name}`;
      if (pkg.local && await pathExists(pkg.local)) {
        logger.debug(`Local ${pkg.name} package found`);
        await copy(pkg.local, targetPath);
        return;
      }
      const cachePath = `${CACHE_DIR}/${pkg.name}`;
      if (await pathExists(`${cachePath}`)) {
        logger.debug(`Cache ${pkg.name} package found`);
        await copy(cachePath, targetPath);
        return;
      }
      logger.info(`No local ${pkg.name} found, we recommend you install ${pkg.name} package locally.`);
      await saveResource(pkg.remote, `${CACHE_DIR}/${pkg.name}`);
      await copy(cachePath, targetPath);
    }));
  }
  async bundleTests() {
    const testDirs = toArray(this.ctx.test.entries);
    const entryPoints = (await Promise.all(testDirs.map((dir) => glob(`${dir}/**/*.spec.{js,ts}`)))).flat();
    this.esbuildContext = await context({
      entryPoints,
      outdir: `${TESTER_PLUGIN_DIR}/content/units`,
      bundle: true,
      target: "firefox115",
      metafile: true
    });
    await this.esbuildContext.rebuild();
  }
  async createTestHtml(tests = []) {
    const setupCode = generateMochaSetup({
      timeout: this.ctx.test.mocha.timeout,
      port: this.port,
      abortOnFail: this.ctx.test.abortOnFail,
      exitOnFinish: !this.ctx.test.watch
    });
    let testFiles = tests;
    if (testFiles.length === 0) {
      testFiles = (await glob(`**/*.spec.js`, { cwd: `${TESTER_PLUGIN_DIR}/content` })).sort();
    }
    const html = generateHtml(setupCode, testFiles);
    await outputFile(`${TESTER_PLUGIN_DIR}/content/index.xhtml`, html);
  }
}
function findImpactedTests(changedFilePath, buildMetadata) {
  if (!buildMetadata)
    return [];
  const resolvedChangedFile = resolve(changedFilePath);
  const impactedTestFiles = /* @__PURE__ */ new Set();
  for (const [outputFilePath, outputInfo] of Object.entries(buildMetadata.outputs)) {
    const testFilePath = outputFilePath.replace(`${TESTER_PLUGIN_DIR}/content/`, "");
    if (Object.keys(outputInfo.inputs).some((inputPath) => resolve(inputPath) === resolvedChangedFile)) {
      impactedTestFiles.add(testFilePath);
    }
  }
  return Array.from(impactedTestFiles);
}

class Test extends Base {
  builder;
  zotero;
  reporter = new TestHttpReporter();
  testBundler;
  constructor(ctx) {
    super(ctx);
    process.env.NODE_ENV ??= "test";
    this.builder = new Build(ctx);
    if (isCI) {
      this.ctx.test.headless = true;
      this.ctx.test.watch = false;
    }
  }
  async run() {
    process.on("SIGINT", this.exit);
    await emptyDir(TESTER_PROFILE_DIR);
    await emptyDir(TESTER_DATA_DIR);
    await emptyDir(TESTER_PLUGIN_DIR);
    await this.ctx.hooks.callHook("test:init", this.ctx);
    await this.builder.run();
    await this.ctx.hooks.callHook("test:prebuild", this.ctx);
    await this.reporter.start();
    this.testBundler = new TestBundler(
      this.ctx,
      this.reporter.port
    );
    await this.testBundler.generate();
    await this.ctx.hooks.callHook("test:bundleTests", this.ctx);
    await this.startZotero();
    await this.ctx.hooks.callHook("test:run", this.ctx);
    if (this.ctx.test.watch) {
      this.watch();
    }
  }
  async watch() {
    const source = toArray(this.ctx.source).map((p) => resolve(p));
    const tests = toArray(this.ctx.test.entries).map((p) => resolve(p));
    function isSource(_path) {
      const path = resolve(_path);
      const isSource2 = source.find((s) => path.match(s)) || false;
      tests.find((t) => path.match(t)) || false;
      return isSource2;
    }
    watch(
      [this.ctx.source, this.ctx.test.entries].flat(),
      {
        onChange: async (path) => {
          if (isSource(path)) {
            await this.builder.run();
            await this.testBundler?.regenerate(path);
            await this.zotero?.reloadAllPlugins();
          } else {
            await this.testBundler?.regenerate(path);
            await this.zotero?.reloadTemporaryPluginBySourceDir(TESTER_PLUGIN_DIR);
          }
        },
        onAdd: async (path) => {
          if (isSource(path)) {
            await this.builder.run();
            await this.testBundler?.regenerate(path);
            await this.zotero?.reloadAllPlugins();
          } else {
            await this.testBundler?.generate();
            await this.zotero?.reloadTemporaryPluginBySourceDir(TESTER_PLUGIN_DIR);
          }
        }
      }
    );
  }
  async startZotero() {
    if (this.ctx.test.headless) {
      await prepareHeadless();
    }
    this.zotero = new ZoteroRunner({
      binary: {
        path: this.zoteroBinPath,
        devtools: this.ctx.server.devtools,
        args: this.ctx.server.startArgs
      },
      profile: {
        path: TESTER_PROFILE_DIR,
        dataDir: TESTER_DATA_DIR,
        customPrefs: this.prefs
      },
      plugins: {
        list: [{
          id: this.ctx.id,
          sourceDir: join(this.ctx.dist, "addon")
        }, {
          id: TESTER_PLUGIN_ID,
          sourceDir: TESTER_PLUGIN_DIR
        }]
      }
    });
    await this.zotero.run();
    this.zotero.zotero?.on("close", () => this.onZoteroExit());
  }
  onZoteroExit = () => {
    this.reporter.stop();
    this.ctx.hooks.callHook("test:exit", this.ctx);
    if (this.reporter.failed)
      process.exit(1);
    else
      process.exit(0);
  };
  exit = (code) => {
    if (code === "SIGINT") {
      this.logger.info("Tester shutdown by user request");
    }
    this.reporter.stop();
    this.zotero?.exit();
    this.ctx.hooks.callHook("test:exit", this.ctx);
    process.exit();
  };
  get zoteroBinPath() {
    if (!process.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH)
      throw new Error("No Zotero Found.");
    return process.env.ZOTERO_PLUGIN_ZOTERO_BIN_PATH;
  }
  get prefs() {
    const defaultPref = {
      "extensions.experiments.enabled": true,
      "extensions.autoDisableScopes": 0,
      // Enable remote-debugging
      "devtools.debugger.remote-enabled": true,
      "devtools.debugger.remote-websocket": true,
      "devtools.debugger.prompt-connection": false,
      // Inherit the default test settings from Zotero
      "app.update.enabled": false,
      "extensions.zotero.sync.server.compressData": false,
      "extensions.zotero.automaticScraperUpdates": false,
      "extensions.zotero.debug.log": 5,
      "extensions.zotero.debug.level": 5,
      "extensions.zotero.debug.time": 5,
      "extensions.zotero.firstRun.skipFirefoxProfileAccessCheck": true,
      "extensions.zotero.firstRunGuidance": false,
      "extensions.zotero.firstRun2": false,
      "extensions.zotero.reportTranslationFailure": false,
      "extensions.zotero.httpServer.enabled": true,
      "extensions.zotero.httpServer.port": 23124,
      "extensions.zotero.httpServer.localAPI.enabled": true,
      "extensions.zotero.backup.numBackups": 0,
      "extensions.zotero.sync.autoSync": false,
      "extensions.zoteroMacWordIntegration.installed": true,
      "extensions.zoteroMacWordIntegration.skipInstallation": true,
      "extensions.zoteroWinWordIntegration.skipInstallation": true,
      "extensions.zoteroOpenOfficeIntegration.skipInstallation": true
    };
    return Object.assign(defaultPref, this.ctx.test.prefs || {});
  }
}

const Config = {
  defineConfig,
  loadConfig
};

export { Build as B, Config as C, ExitSignals as E, Release as R, Serve as S, Test as T, defineConfig as d, logger as l };
